### PR TITLE
regexp-condition-v2: add tests for NULL pattern with indexscan and bitmapscan

### PR DESCRIPTION
**Pull request by komainu8**

### CODE REVIEW

**Comment by kou** [comment]
indexscan/bitmapscanとseqscanで処理が違うから他のと同じように3パターンのテストをして。

**Comment by komainu8** [comment]
> indexscan/bitmapscanとseqscanで処理が違うから他のと同じように3パターンのテストをして。

承知です。

**Comment by komainu8** [comment]
`EXPLAIN`もエラーになるのか。

```
CREATE TABLE memos (
  content text[]
);

INSERT INTO memos
     VALUES (ARRAY['PostgreSQL is a RDBMS',
                   'Groonga is fast full text search engine',
                   'PGroonga is a PostgreSQL extension that uses Groonga']);

INSERT INTO memos
     VALUES (ARRAY['MySQL is a RDBMS',
                   'Mroonga is a MySQL storage engine that uses Groonga']);

CREATE INDEX pgrn_content_index
    ON memos
 USING pgroonga (content pgroonga_text_array_regexp_ops_v2);

SET enable_seqscan = off;
SET enable_indexscan = on;
SET enable_bitmapscan = off;

EXPLAIN (COSTS OFF)
SELECT *
  FROM memos
 WHERE content &~ pgroonga_condition(NULL,
                                     index_name => 'pgrn_content_index');
ERROR:  pgroonga: [build-condition][regexp-condition] query must not NULL
```

**Comment by komainu8** [comment]
コストの推定の時に`PGrnSearchBuildConditionBinaryOperationCondition()`が呼ばれるから`EXPLAIN`でもエラーになるのか。

**Comment by komainu8** [comment]
`indexscan`のプランのコスト推定と、`bitmapscan`のプランのコスト推定と、`seqscan`のプランのコスト推定でそれぞれ処理が違うから、それぞれのテストを作って、`EXPLAIN`も普通の検索もちゃんとエラーで弾かれるか確認できればいいのか。

**Comment by kou** [comment]
いや、そうじゃなくて、`indexscan`/`bitmapscan`は`EXPLAIN`でエラーでいいけど、`seqscan`は実行時にエラーになって欲しい。

**Comment by kou** [comment]
この話、今回の一連の作業の中で何回か同じことを言っている気がするんですけど、もしかして、まだピンときていません？

**Comment by komainu8** [comment]
うーん。そうですね。わかってないですね。

`NULL`の判定は`PGrnSearchBuildConditionPrepareCondition()`でやってますが、`seqscan`の時は`EXPLAIN`で`PGrnSearchBuildConditionPrepareCondition()`が呼ばれず、実行時には呼ばれる理由がよくわかってないです。

**Comment by askdkc** [comment]
今日のグルカイネタの誕生か💡

**Comment by kou** [comment]
`seqscan`では実行時に`PGrnSearchBuildConditionPrepareCondition()`は呼ばれません。`EXPLAIN`のときにしか呼ばれる可能性はないです。

今のテストでPGrnSearchBuildConditionPrepareCondition()`が呼ばれているのはインデックスを作っているからです。
https://github.com/pgroonga/pgroonga/pull/592/files#diff-44bd87cf07b0c3e82248c03369f15c5e09024ecee2f7618eac3a2b1e9204f535R14-R16

これを消せば`seqscan`のときには`EXPLAIN`のときも実行時も呼ばれません。

**Comment by komainu8** [comment]
あぁ、最終的に`seqscan`が選択される場合であっても、PostgreSQLは実行可能なプランそれぞれについてコストを計算するので、インデックスが存在すると、それを使った検索のコストも計算する。
だから、今のテストだと`PGrnSearchBuildConditionPrepareCondition()`が呼ばれるということですね。

**Comment by komainu8** [comment]
そして、それで実行すると、エラーにならないので、`seqscan`のときの`NULL`チェックを追加しないといけない。

```
SELECT *
   FROM memos
WHERE content &~ pgroonga_condition(NULL);
-- content 
-- ---------
-- (0 rows)
```

**Comment by komainu8** [comment]
`indexsca`と`bitmapscan`も`index_name`の指定はいらない。

**Comment by komainu8** [comment]
`seqscan`のときの`NULL`チェックと`seqscan`のテストは #593 へ分離した。
このPRは`indexscan`と`bitmapscan`のテスト追加のみにする。

File: sql/regexp/text-array/regexp-condition-v2/null/bitmapscan.sql [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+SET enable_bitmapscan = on;
+
+EXPLAIN (COSTS OFF)
+SELECT *
+  FROM memos
+ WHERE content &~ pgroonga_condition(NULL);
+
+SELECT *
+  FROM memos
+ WHERE content &~ pgroonga_condition(NULL);
```
Comment (kou) [inline sql/regexp/text-array/regexp-condition-v2/null/bitmapscan.sql:29]: `EXPLAIN`でエラーになるから`EXPLAIN`なしバージョンはいらないです。

File: expected/regexp/text-array/regexp-condition-v2/null/bitmapscan.out [hunk 2]
```diff
@@ -0,0 +1,26 @@
+CREATE TABLE memos (
+  content text[]
```
Comment (kou) [inline expected/regexp/text-array/regexp-condition-v2/null/bitmapscan.out:2]: `contents`にして。

File: sql/regexp/text-array/regexp-condition-v2/null/bitmapscan.sql (hunk 1 above)
Comment (komainu8) [reply sql/regexp/text-array/regexp-condition-v2/null/bitmapscan.sql]: `EXPLAIN`なしバージョンを削除しました。

File: expected/regexp/text-array/regexp-condition-v2/null/bitmapscan.out (hunk 2 above)
Comment (komainu8) [reply expected/regexp/text-array/regexp-condition-v2/null/bitmapscan.out]: `contents`にしました。
